### PR TITLE
Decision Trees and Debugging

### PR DESCRIPTION
**Pull request by KilnOfTheSecondFlame**

### CODE REVIEW

File: FS19/ML/Zusammenfassung_ML.tex [hunk 1]
```diff
@@ -2370,6 +2370,114 @@ \subsection{Evaluation}
 
 An often used performance measurement is $P@k$ (Precision @ k). How many of the $k$ recommended items were actually relevant. In case many users have less than $k$ recommended products, you could also go with $minP@k$.
 
+\section{Decision Trees and Random Forests}
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:2376]: Add short definition/explanation what decision trees even are and what they're used for

File: FS19/ML/Zusammenfassung_ML.tex [hunk 2]
```diff
@@ -2370,6 +2370,114 @@ \subsection{Evaluation}
...
+    \label{fig:decisiontreeexample}
+\end{figure}
+
+\subsection{Tree Construction Rules}
+
+\begin{enumerate}
+    \item If only positive or negative instances are left, stop branching and assign the corresponding decision as leaf node\\Example: On Offer = 0\% or 50\%
+    \item If some positive and some negative examples remain, choose another feature and refine the tree one step further\\Example: On Offer = 20\% $\rightarrow$ next feature: Saturday
+    \item If no instances are left, such a combination of feature values does not occur in the training set. In this case, look at the parent node and decide according to the more frequent label
+    \item If there are still instances but no features any more, the training data s contradictory, noisy, non-deterministic or contains hidden features. In his case, decide according to the more frequent label.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: [...] the training data _is_ contradictory [...]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 3]
```diff
@@ -2370,6 +2370,114 @@ \subsection{Evaluation}
...
+\end{enumerate}
+
+\subsubsection{Splitting Criterion}
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:2406]: Splitting Criteria?

File: FS19/ML/Zusammenfassung_ML.tex [hunk 4]
```diff
@@ -2370,6 +2370,114 @@ \subsection{Evaluation}
...
+    \item[$p(\text{Weather = Cloudy})$] $\frac{2}{12}$
+\end{itemize}
+
+\begin{equation*}
+    \text{Impurity(Weather)} = 1 - (\frac{7}{12})^2 - (\frac{3}{12})^2 - (\frac{2}{12})^2 = \frac{41}{72} \approx 0.569
+\end{equation*}
+
+\subsection{Regression Trees in CART}
+
+The regression tress predict the average value of all instances in a leaf node and are grown like classification trees but with a different splitting criteria. As splitting criteria, they minimise the variance of the values in the same subset.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: The regression _trees_ [...]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 5]
```diff
@@ -2370,6 +2370,114 @@ \subsection{Evaluation}
...
+
+\subsection{Advantages and Disadvantages of Decision Trees}
+
+\textbf{Advantages}
+\begin{itemize}
+    \item Decision trees are simple to understand and interpret for humans
+    \item Decision trees can handle both numeric and categorical data
+    \item Decision trees require almost no data preparation
+    \item Decision trees perform well even on larger datasets
+\end{itemize}
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:2446]: - Simple to understand [...]
- Can handle both [...]
- Require almost no [...]
- Perform well [...]

Leave the Decision tree out.

File: FS19/ML/Zusammenfassung_ML.tex [hunk 6]
```diff
@@ -2370,6 +2370,114 @@ \subsection{Evaluation}
...
+\textbf{Disadvantages}
+\begin{itemize}
+    \item Trees are often not as accurate as other approaches
+    \item Learners tend to create over-complex trees that do not generalize well - they overfit the data
+    \item A small change in the training data can result in a big change in the tree - highly sensitive to the training data
+\end{itemize}
+
+\subsection{Association Rules versus Decision Trees}
+
+Mining association rules is unsupervised learning, whereas building decision trees is supervised learning. Where association rules process pure transactions, decision trees incorporate other information, therefore both techniques can be used in a personalized and impersonalized manner.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: Whereas association rules [...]

Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: [...] incorporate other information. Therefore, both [...]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 7]
```diff
@@ -2370,6 +2370,114 @@ \subsection{Evaluation}
...
+
+\subsection{Random Forest}
+
+Random forests try to address the problems of decision trees. They are simply a collection of decision trees built from the same training set using \textbf{random sampling with replacement} and a \textbf{ random attribute selection}, the prediction from a forest is a combination of the individual tree's results.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: [...] and a \textbf{random attribute selection}. The prediction [...]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 8]
```diff
@@ -2480,4 +2482,123 @@ \subsubsection{Building a Random Forest}
...
+    \item This model performs badly on both training and test data
+\end{itemize}
+
+\noindent
+\textbf{Overfitting}
+\begin{itemize}
+    \item This model optimizes to much with respect to the training data
+    \item It shows high performance on training but low performance on test data
+\end{itemize}
+
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:2546]: Do over- and underfitting in regression and classification really need to be separate if they're basically the same anyway?

File: FS19/ML/Zusammenfassung_ML.tex [hunk 9]
```diff
@@ -2480,4 +2482,123 @@ \subsubsection{Building a Random Forest}
...
+\begin{figure}[h!]
+    \centering
+    \includegraphics[width=0.7\linewidth, keepaspectratio]{Pictures/overfitting_learning_curve}
+    \caption{Overfitting has good performance on training but bad performance on test data}
+    \label{fig:overfittinglearningcurve}
+\end{figure}
+
+\subsubsection{Counteracting Overfitting}
+Overfitting happens more frequently and is much harder to repair. It occurs cause your model is to sensitive to the training data, you could try to:
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: It occurs _'cause_ your model [...] 
or
It occurs _because_ your model [...]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 10]
```diff
@@ -2480,4 +2482,123 @@ \subsubsection{Building a Random Forest}
...
+\begin{itemize}
+    \item Try getting more training data
+    \item Try better data cleaning and eliminate outliers
+    \item Try a smaller set of features through feature selection
+    \item Try early stopping during iterative training
+    \item Try tuning the regularization parameter
+    \item Try ensemble methods
+\end{itemize}
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:2584]: Does it need a 'try' before every item?

File: FS19/ML/Zusammenfassung_ML.tex [hunk 11]
```diff
@@ -2480,4 +2482,123 @@ \subsubsection{Building a Random Forest}
...
+    \label{fig:earlystopping}
+\end{figure}
+
+\subsection{How to Approach a Machine Learning Project}
+
+\begin{enumerate}
+    \item Data quality assessment is always worth the time
+    \item Visualize your data
+    \item Specify and implement a water-proof quality assessment workflow
+    \item Implement a quick-and-dirty baseline classifier or regressor
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:2601]: Add the example with the baseline classifier from the lecture

**Review by gathurian — CHANGES_REQUESTED** [review]

File: FS19/ML/Zusammenfassung_ML.tex (hunk 3 above)
Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex:2406]: grafik

File: FS19/ML/Zusammenfassung_ML.tex (hunk 8 above)
Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex:2546]: But it's not the same thing =/
One you are looking for a line to separate classes, another time to fit it to your data points

File: FS19/ML/Zusammenfassung_ML.tex (hunk 11 above)
Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex:2601]: I did not write that down

**Review by gathurian — APPROVED** [review]

**Pull request merged**